Sources/Sentry/SentrySampling.m

```diff
@@ -0,0 +1,112 @@
+#import "SentrySampling.h"
+#import "SentryDependencyContainer.h"
+#import "SentryOptions.h"
+#import "SentryRandom.h"
+#import "SentrySampleDecision.h"
+#import "SentrySamplerDecision.h"
+#import "SentrySamplingContext.h"
+#import "SentryTransactionContext.h"
+#import <SentryOptions+Private.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+#pragma mark - Private
+
+/**
+ * @return A sample rate if the specified sampler callback was defined on @c SentryOptions and
+ * returned a valid value, @c nil otherwise.
+ */
+NSNumber *_Nullable samplerCallbackRate(
+    SentryTracesSamplerCallback callback, SentrySamplingContext *context, SentryOptions *options)
+{
+    if (callback == nil) {
+        return nil;
+    }
+
+    NSNumber *callbackRate = callback(context);
+    if (callbackRate == nil || !isValidSampleRate(callbackRate)) {
```

Comment:
`h`: The `callback` returns=ing `nil` means no sampling decision has been taken, if I'm not mistaken, and we can't fallback to the defaultSampleRate. I think this would be a change in behaviour to what we previously had see https://github.com/getsentry/sentry-cocoa/blob/50bb751cfe50189673d3c0ec5b0d664659218ade/Sources/Sentry/SentryTracesSampler.m#L50-L60

Adding these two test, which represent the old behaviour,  to the `SentryHubTests` should clarify if this breaks something or not
```swift
    func testStartTransaction_WhenSamplerNil_NotSampled() {
        SentryDependencyContainer.sharedInstance().random = TestRandom(value: 0.5)
        assertSampler(expected: .no) { options in
            options.tracesSampleRate = 0.49
            options.tracesSampler = { _ in return nil }
        }
    }
    
    func testStartTransaction_WhenSamplerNil_Sampled() {
        SentryDependencyContainer.sharedInstance().random = TestRandom(value: 0.5)
        assertSampler(expected: .yes) { options in
            options.tracesSampleRate = 0.50
            options.tracesSampler = { _ in return nil }
        }
    }
```

Reply:
You're correct I didn't transcribe this code correctly. But looking again, I don't see anywhere `defaultTracesSampleRate` is ever set to anything other than `nil`, and it's not a public property. Looks like it can just be removed.

Reply:
Also, I see now that we're using `nil` as a stand-in for the default traces sample rate of `0`. I disagree with `tracesSampleRate` ever having been made nullable (I dug up this comment thread: https://github.com/getsentry/sentry-cocoa/pull/961#discussion_r582730070–I see you agree with me here). Unfortunately making the properties and `SentryTracesSamplerCallback` nonnull would be a breaking change at this point. I'll make a personal note to recommend changing it in the next major.

What I'm going to do now is remove the `@property` declarations and make the definitions of the defaults a preprocessor constant. No need for these to be part of the instance.

Reply:
For this reason also, the check for whether or not the sampler function returns nil never made sense; we're just returning the same thing anyways, and also, it evaluates to a valid sample rate, since 0 is valid. ¯\_(ツ)_/¯

Reply:
> I'll make a personal note to recommend changing it in the next major.

👍 to ⬆️ 

Yeah, it seems like you didn't transcribe this code correctly, which made me suspicious, but it was how the old code works. Kind of weird 😄

Reply:
So I ironed out a bit of the weirdness with `nil`, and added those tests, which both still pass. Thanks for sharing those.

Reply:
https://github.com/getsentry/sentry-cocoa/discussions/2946?sort=new#discussioncomment-9088191